### ISSUE TITLE
Page.publisher_public_id sometimes loses it's reference

### ISSUE THREAD

**Issue by mikemcgowan** [django-cms/django-cms#4940]
Our production systems are currently on Python 2.7, Django 1.7, CMS 3.1, MySQL 5.6. (We are in the process of moving to Django 1.8, CMS 3.2.)

Occasionally (at least 3 times in the last 6 months) publishing a page causes the `cms_page.publisher_public_id` foreign key to get set to null on either _a)_ the draft version or _b)_ both the draft and non-draft versions of the 'pair' of `cms_page` records. This corrupt data can then result in several unhandled exceptions, including attempting to invoke `cms.api.get_page_draft` on a non-draft page:

``` python
def get_page_draft(page):
    """
    Returns the draft version of a page, regardless if the passed in
    page is a published version or a draft version.

    :param page: The page to get the draft version
    :type page: :class:`cms.models.pagemodel.Page` instance
    :return page: draft version of the page
    :type page: :class:`cms.models.pagemodel.Page` instance
    """
    if page:
        if page.publisher_is_draft:
            return page
        else:
            return page.publisher_draft
    else:
        return None
```

Specifically, when invoked on a non-draft page, the line `return page.publisher_draft` throws an exception because the reverse lookup fails due to the foreign key being null.

This issue is intermittent and we cannot reproduce the problem, so unfortunately, I'm unable to run it through a debugger to get any better information.

**Comment by mikemcgowan** [comment]
See also https://github.com/divio/django-cms/issues/3274

**Comment by mikemcgowan** [comment]
Below is a Django management command that fixes the null/lost/corrupt `cms_page.publisher_public_id` columns. It works on the assumption that the equivalent columns in `cms_title` are okay. In our experience, it's only the `cms_page.publisher_public_id` columns that get broken, so this assumption holds for us.

```python
from django.core.management.base import BaseCommand

from cms.models import Title


class Command(BaseCommand):
    help = 'Fix CMS pages (publisher_public_id foreign key)'

    def handle(self, *args, **options):
        ts = Title.objects.filter(
            publisher_is_draft=False,
            page_id__isnull=False,
            publisher_public_id__isnull=False
        )

        fixed_count = 0

        # for each of the (non-draft, draft) title pairs, fix the equivalent (non-draft, draft) page pairs
        for t in ts:
            fixed = False

            if t.publisher_public.page.publisher_public is None:
                self.stdout.write('Draft page id {} has a null publisher_public_id (it should be {})'.format(
                    t.publisher_public.page_id,
                    t.page_id
                ))
                t.publisher_public.page.publisher_public = t.page
                t.publisher_public.page.save()
                fixed = True

            if t.page.publisher_public is None:
                self.stdout.write('Non-draft page id {} has a null publisher_public_id (it should be {})'.format(
                    t.page_id,
                    t.publisher_public.page_id
                ))
                t.page.publisher_public = t.publisher_public.page
                t.page.save()
                fixed = True

            if fixed:
                fixed_count += 1

        self.stdout.write('Fixed {} pairs of pages.'.format(fixed_count))
```

**Comment by mikemcgowan** [comment]
Here is part of a stack trace that could be helpful:

```
[19/Jan/2016 10:05:30] ERROR [django.request:231] Internal Server Error: /en/
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/cms/views.py", line 188, in details
    context['has_view_permissions'] = page.has_view_permission(request)
  File "/usr/local/lib/python2.7/dist-packages/cms/models/pagemodel.py", line 1029, in has_view_permission
    is_restricted = bool(get_any_page_view_permissions(request, self))
  File "/usr/local/lib/python2.7/dist-packages/cms/utils/permissions.py", line 336, in get_any_page_view_permissions
    page = page.publisher_draft
  File "/usr/local/lib/python2.7/dist-packages/django/db/models/fields/related.py", line 428, in __get__
    self.related.get_accessor_name()
RelatedObjectDoesNotExist: Page has no publisher_draft.
```

**Comment by vishnuratheesh** [comment]
I experienced the same on Django 1.8.9 / Django-CMS 3.2.0 / Postgres 9.1

The management command helped! Thanks a bunch @mikemcgowan

**Comment by imagescape** [comment]
Same situation. The management command worked like a charm.  Thanks for sharing. 

Django==1.8.9
django-cms==3.2.0
django-reversion==1.9.3
Postgres

**Comment by yakky** [comment]
django CMS 3.3 and 3.4 contains a lot of fixes regarding the tree handling. Some of them have been backported to 3.2.5. You are encouraged to upgrade to fix those

**Comment by czpython** [comment]
Hello @bartTC,
I don't think we've made modifications in regards to transactions on 3.4, at least no on the publication of pages.

Using your script, combined with cancelling, are you able to reproduce the issue on 3.4?

**Comment by bartTC** [comment]
I am able to reproduce this within a regular production build by triggering a reversion. That one fails (I dont know the reason yet) with the standard exception. 

> RevertError: Could not revert revision, due to database integrity errors.

The result is a lost Public > Draft relation, it can be fixed by the script(s) mentioned here. 

I'm trying to encapsulate this into a test next.

**Comment by czpython** [comment]
@bartTC revisions were removed on cms 3.4 (this was one of the reasons), are you using an earlier version?

**Comment by Aiky30** [comment]
Closing as this issue appears to be stale and there is no feedback that the proposed fixes in 3.4 did not work. I have never heard of this on any project 3.5+ so assuming fixed. 

Due to the age of this issue we are unable to safely accept the contribution. If you feel this issue is still required please inform a member of our technical committee: [nicolai@django-cms.org](mailto:nicolai@django-cms.org ).

Nevertheless we thank you for your participation and would like to emphasize that we do not want to discourage you in your contribution to the django CMS project. On the contrary, we encourage everyone who is interested in the success of django CMS to join us!

If you want to learn more about the work of the technical committee, our product vision and how you can become a member of the dCA, you can find all information here:

- [The Technical Committee of the django CMS Association](https://github.com/django-cms/django-cms-mgmt/blob/master/tech-committee/about.md)
- [Technical Roadmap and Product Vision for django CMS](https://github.com/django-cms/django-cms-mgmt/blob/master/django-cms/roadmap.md)
- [How to become a member of the django CMS Association](https://github.com/django-cms/django-cms-mgmt/blob/master/association/about.md)

We are looking forward to every companion who would like to shape the future of django CMS together with us! 

If you have any questions, feel free to reach out to Nicolai, our Community Manager: nicolai@django-cms.org 


Sign up for the Association now!
